### DISCUSSION TITLE
[Question] What debugger should I use to make cheats?

### DISCUSSION THREAD

**Discussion by SirEnder125** [RedShyGuy/Vapecord-ACNL-Plugin#24] [Q&A]
~Hello. I tried downloading GHIDRA debugger, though, it came as a
source code. I have no idea how to compile it. Any helpful suggestions?~

I built it (I think) but now I don't know how to use it....
@RedShyGuy do you know how to use it?
Also, it DOES have a UI.

**Comment by RedShyGuy** [comment]
Ha maybe there came a new version I didn't see lol. Sadly i don't know how to use it as I use IDA. But I can check it out tomorrow!

**Reply by SirEnder125** [reply]
Have you tried it yet? (No rush lol)

**Reply by RedShyGuy** [reply]
No sorry!!
I meant I can test it on the 24th. I missaid that I can test it today sorry 😶.

**Reply by SirEnder125** [reply]
Oh, Ok. Understood. I made an account on gbatemp to ask someone about
ghidra and their reply sounded like they thought my question was very dumb...
So I'm not sure I can even use it for debugging 3ds.

**Reply by RedShyGuy** [reply]
You can use it for that, it was just not directly designed for the 3DS.

**Reply by SirEnder125** [reply]
Oh, well look at the replies to this:
https://gbatemp.net/threads/using-ghidra-to-debug-3ds-games.579402/#post-9306736

I have heard of something called "gdb", some kind of debugger... I can't use it though.

**Reply by RedShyGuy** [reply]
Of course ghidra itself is not a debugger lol, I never said that. But it has a gdb debugger.

**Reply by SirEnder125** [reply]
Oh, ok. Sorry, I was confused. 🤣

**Reply by SirEnder125** [reply]
Have you tried it?

**Reply by SirEnder125** [reply]
Ok, I guess I'll try to figure it out myself.
Not going to be easy...

**Comment by SirEnder125** [comment]
Is there for real only 1 debugger that will work?

**Reply by RedShyGuy** [reply]
I only know of Ghidra and IDA

**Reply by SirEnder125** [reply]
Have you tested ghidra?

**Reply by SirEnder125** [reply]
By the way, do you know the address to unlock the shampoodle?

**Reply by RedShyGuy** [reply]
No I did not test it yet sorry.
And I do know the address, I can give it to you later.

**Reply by SirEnder125** [reply]
Ok, any time. Sorry for bugging you. Just trying to find out
how to DEbug the game... (Bad pun, I know.) 😜

**Reply by RedShyGuy** [reply]
Soo the address for unlocking shampoodle is:
```cpp
u32 GardenPlus = 0x31F26F80; //Beginning of Garden_plus.dat

u32 Shampoodle = GardenPlus + 0x6ADB4;

Process::Write8(Shampoodle, 2);
```

**Reply by SirEnder125** [reply]
Thank you very much! You've been a huge help. :D

**Reply by SirEnder125** [reply]
By the way, by what means did you learn ASM? I want to learn some ASM
and test with it.

**Reply by RedShyGuy** [reply]
I just looked up all the stuff I actually need, I dont know very much about asm. Only really important stuff😅

**Reply by SirEnder125** [reply]
Ok, got it... I wouldn't know what keywords to search but that isn't my greatest problem.
I need a debugger!! :/

**Reply by SirEnder125** [reply]
I found a link that talks a bit about GBA rom debugging with Ghidra, but the installation
steps fail every time...

**Comment by SirEnder125** [comment]
Okay, after watching a tutorial, I finally got Ghidra running. If anyone could help with
3ds debugging, that'd be great. :wink:

**Reply by SirEnder125** [reply]
Okay, I don't understand why it could be used to debug a 3ds.
Being developed by the NSA, do they really have a need to debug
the 3ds? I don't think so. If, however, it can debug, please tell me.

**Reply by RedShyGuy** [reply]
It is not specifcly designed to debug a 3ds lol. It just supports GDB debugging. Which the 3ds also supports if you have Luma CFW

**Reply by SirEnder125** [reply]
I wouldn't think so, lol. I didn't find an actual option to debug or anything,
do you have ghidra installed?

**Reply by RedShyGuy** [reply]
Yea I do, and I did use the debugger once.

**Reply by SirEnder125** [reply]
Oh, cool! Did it work well? How can I use the debugger?
Also, where is it? Is it a built-in function?

**Reply by SirEnder125** [reply]
Sorry, I know this is a lot to ask.

**Reply by RedShyGuy** [reply]
It did work well, but as it had no interface at all I didn't use it much, IDA just works much better. I can't really remember where I found it to be honest. May I ask, is there any way I can privately contact you?

**Reply by SirEnder125** [reply]
Do you have Skype?

**Reply by RedShyGuy** [reply]
No sorry. But I have Twitter, Reddit, Instagram, Tumblr, Discord. Many ways to reach me😜

**Reply by SirEnder125** [reply]
Well, I don't have any of those. :/ oof.

**Reply by SirEnder125** [reply]
I hear that Ghidra 9.3 (or whatever version is next) will have a debugger.
Is this what you tried? I would think it to have a gui, Ghidra does.

**Reply by RedShyGuy** [reply]
I do remember that the debugger was a standalone app that came with Ghidra, though I can't remember where it was. You should really just get IDA, its much better and works very well.

**Reply by SirEnder125** [reply]
I'm not saying you did, but I know that most plugin makers get their IDA pro
off of "piratebay" (and/or some other pirating website). I will not do that. It is
wrong. I will just have to not use a debugger if that is my only other choice.

**Reply by SirEnder125** [reply]
I actually just built the ghidra-debugger branch and tried to connect to my
2ds. On any game it crashes.... :/ Even the home screen

**Reply by RedShyGuy** [reply]
Oof :/

**Reply by SirEnder125** [reply]
I have tried connecting with arm-none-eabi-gdb.exe but the same thing happens. Is it my console that causes this?

EDIT: I tried it on my second 2DS and the same thing happens. :/

**Reply by SirEnder125** [reply]
Hey, I have a question:
What was the best cheat you've made without a debugger?
(Or what can I make with only GDB as a debugger)

**Reply by RedShyGuy** [reply]
Save Edit Cheats! You don't need any debugger for save edits as they are easy to find

**Reply by SirEnder125** [reply]
Oh, okay. Thanks.
